### ISSUE TITLE
Mismatch between regular string and interpolated string

### ISSUE THREAD

**Issue by futile** [nushell/nushell#11069]
### Describe the bug

Given the following two files `bug.nu` and `dump-args.sh`:

`bug.nu`:
```nushell
#!/usr/bin/env nu

def main [] {
  ^./dump-args.sh '"foobar"'

  let s = "foobar"
  ^./dump-args.sh $'"($s)"'
}
```

`dump-args.sh`:
```bash
#!/usr/bin/env bash

echo "Arguments: $@"
```

Running `./bug.nu` outputs:
```console
❯ ./bug.nu
Arguments: "foobar"
Arguments: foobar
```

So the double-quotes (`"`) from `'"foobar"'` are not present in `$'"($s)"'`.

Workaround: Doubling the double-quotes helps somehow, so this:
```nu
^./dump-args.sh $'""($s)""'
```
correctly prints out `"foobar"` again.

### How to reproduce

See the description, it contains two files to reproduce.

### Expected behavior

I expected `'"foobar"'` to be the same as `let s = "foobar"; $'"($s)"'`, but the latter lost the double-quotes.

### Configuration

`❯ nu -c "version | transpose key value | to md --pretty"`
| key                | value                                                             |
| ------------------ | ----------------------------------------------------------------- |
| version            | 0.85.0                                                            |
| branch             |                                                                   |
| commit_hash        |                                                                   |
| build_os           | linux-x86_64                                                      |
| build_target       | x86_64-unknown-linux-gnu                                          |
| rust_version       | rustc 1.72.0 (5680fa18f 2023-08-23) (built from a source tarball) |
| cargo_version      | cargo 1.72.0                                                      |
| build_time         | 1980-01-01 00:00:00 +00:00                                        |
| build_rust_channel | release                                                           |
| allocator          | mimalloc                                                          |
| features           | default, sqlite, trash, which, zip                                |
| installed_plugins  |                                                                   |

**Comment by fdncred** [comment]
Does this still happen in 0.87.0?

**Comment by futile** [comment]
Yup, looks like it still happens (sorry for not checking a more up-to-date version before!):

```console
❯ ./bug.nu
Arguments: "foobar"
Arguments: foobar

❯ nu -c "version | transpose key value | to md --pretty"
| key                | value                                                             |
| ------------------ | ----------------------------------------------------------------- |
| version            | 0.87.0                                                            |
| branch             |                                                                   |
| commit_hash        |                                                                   |
| build_os           | linux-x86_64                                                      |
| build_target       | x86_64-unknown-linux-gnu                                          |
| rust_version       | rustc 1.72.1 (d5c2e9c34 2023-09-13) (built from a source tarball) |
| cargo_version      | cargo 1.72.0                                                      |
| build_time         | 1980-01-01 00:00:00 +00:00                                        |
| build_rust_channel | release                                                           |
| allocator          | mimalloc                                                          |
| features           | default, sqlite, trash, which, zip                                |
| installed_plugins  |                                                                   |
```

**Comment by MarikaChlebowska** [comment]
I think my PR #11030 fixes it but right now I don't have time to try it out

**Comment by futile** [comment]
> I think my PR #11030 fixes it but right now I don't have time to try it out

I just tried it out, here is the result:

```
[foo@bar:~/gits/nu-bug-interpolation-quotes]$ ../marika-nushell/target/debug/nu bug.nu
Arguments: "foobar"
Arguments: foobar

[foo@bar:~/gits/nu-bug-interpolation-quotes]$ ../marika-nushell/target/debug/nu --version
0.86.1

[foo@bar::~/gits/nu-bug-interpolation-quotes]$ nu -c "version | transpose key value | to md --pretty"
```
| key                | value                                                             |
| ------------------ | ----------------------------------------------------------------- |
| version            | 0.86.0                                                            |
| branch             |                                                                   |
| commit_hash        |                                                                   |
| build_os           | linux-x86_64                                                      |
| build_target       | x86_64-unknown-linux-gnu                                          |
| rust_version       | rustc 1.72.1 (d5c2e9c34 2023-09-13) (built from a source tarball) |
| cargo_version      | cargo 1.72.0                                                      |
| build_time         | 1980-01-01 00:00:00 +00:00                                        |
| build_rust_channel | release                                                           |
| allocator          | mimalloc                                                          |
| features           | dataframe, default, extra, sqlite, trash, which, zip              |
| installed_plugins  |                                                                   |

So sadly didn't fix it :( But thanks for checking out & linking it! :)

**Comment by MarikaChlebowska** [comment]
That's a shame. But it's most likely issue in the same file I changed in my PR so when I have more time I can try to figure out what's wrong there

**Comment by futile** [comment]
Thank you! That would be awesome! :)

**Comment by MarikaChlebowska** [comment]
And it's not related to my PR since the path for parsing arguments for nu scripts and bash scripts is not the same. I still want to dig into it so maybe I'll create a PR if I have a solution.

**Comment by MarikaChlebowska** [comment]
As far as I see the issue is with the way string interpolations and variables are passed to external functions. I made a simpler script where the same issue occurs but I'm far from finding a solution for that
```
def main [] {
  let s = '"foobar"'
  ^echo '"foobar"' # prints "foobar"
  ^echo $s # prints foobar
  ^echo $"($s)" # prints foobar
}
```

**Comment by MarikaChlebowska** [comment]
After a long debugging session I found where does the issue occur and possibly fixed it.
TLDR
I changed the way arguments to external calls work and fixed the bug on my branch (it might break other calls):
[<branch link>](https://github.com/MarikaChlebowska/nushell/tree/quote-external-command-arguments)

Long story
Right now external commands (which includes calling bash scripts) don't have their non-dollar arguments parsed at all, meaning calling `./script.sh '"something"'` in nu passes exactly `'"something"'` to bash without processing it in any way. When there are dollar expressions in the arguments they are replaced as the content of the expression in the call without adding external quotes or anything. For example
```
let s = 'string'
^echo $s # this will be like calling `echo string` in bash
let s = '"string"'
^echo $s # this will be like calling `echo "string"` in bash
```
Both of these expressions will have the same result. I created a branch where a FullCellPath or StringInterpolation passed to external commands, if they evaluate to string will have quotes implicitly added. This might break other external calls so it requires extensive testing first so I'll look some more into it before creating a PR.

**Comment by futile** [comment]
Wow, awesome, thank you so much @MarikaChlebowska !! Really appreciate the work & time you put into debugging this!

> Right now external commands (which includes calling bash scripts) don't have their non-dollar arguments parsed at all, meaning calling ./script.sh '"something"' in nu passes exactly '"something"' to bash without processing it in any way. When there are dollar expressions in the arguments they are replaced as the content of the expression in the call without adding external quotes or anything.

That actually makes a lot of sense, at least regarding the behavior that occurred. Whether it should be like this is a different question. But this behavior can at least be documented in the docs, and is then understandable/findable for other users as well.

 Again, thank you so much for digging so deep into this :) <3

**Comment by fdncred** [comment]
From discord,
it looks like the trim is happening here https://github.com/nushell/nushell/blob/15c7e1b725e4bee09d66d1b3af247c98411818a0/crates/nu-command/src/system/run_external.rs#L713 when run-external is called.
This PR seems to remove quotes on purpose https://github.com/nushell/nushell/pull/7740

**Comment by Bahex** [comment]
Works as expected on `0.114.1`, given sample prints:
```
Arguments: "foobar"
Arguments: "foobar"
```
